sale_invoice_policy/models/sale_order_line.py

```diff
@@ -29,74 +25,27 @@ def _compute_qty_to_invoice(self):
-            uom_qty_to_consider = (
-                line.qty_delivered
-                if invoice_policy == "delivery"
-                else line.product_uom_qty
-            )
-            price_reduce = line.price_unit * (1 - (line.discount or 0.0) / 100.0)
-            price_subtotal = price_reduce * uom_qty_to_consider
-            if len(line.tax_id.filtered(lambda tax: tax.price_include)) > 0:
-                price_subtotal = line.tax_id.compute_all(
-                    price_reduce,
-                    currency=line.currency_id,
-                    quantity=uom_qty_to_consider,
-                    product=line.product_id,
-                    partner=line.order_id.partner_shipping_id,
-                )["total_excluded"]
-            inv_lines = line._get_invoice_lines()
-            if any(inv_lines.mapped(lambda l: l.discount != line.discount)):
-                amount = 0
-                for inv_line in inv_lines:
-                    if (
-                        len(inv_line.tax_ids.filtered(lambda tax: tax.price_include))
-                        > 0
-                    ):
-                        amount += inv_line.tax_ids.compute_all(
-                            inv_line.currency_id._convert(
-                                inv_line.price_unit,
-                                line.currency_id,
-                                line.company_id,
-                                inv_line.date or fields.Date.today(),
-                                round=False,
-                            )
-                            * inv_line.quantity
-                        )["total_excluded"]
-                    else:
-                        amount += (
-                            inv_line.currency_id._convert(
-                                inv_line.price_unit,
-                                line.currency_id,
-                                line.company_id,
-                                inv_line.date or fields.Date.today(),
-                                round=False,
-                            )
-                            * inv_line.quantity
-                        )
-                amount_to_invoice = max(price_subtotal - amount, 0)
-            else:
-                amount_to_invoice = price_subtotal - line.untaxed_amount_invoiced
-            line.untaxed_amount_to_invoice = amount_to_invoice
-        return True
+        saved_product_values = dict()
+        lines = self - other_lines
+        for line in lines:
+            # Save product invoice policy, change it to sale order one, add context manager
+            # to not trigger other computes. Then, restablish former one.
+            with self.env.norecompute():
+                saved_product_values[line.product_id] = line.product_id.invoice_policy
+                line.product_id.invoice_policy = line.order_id.invoice_policy
+            # We should call the super() per line as each sale order can have different policy
```

Comment:
To avoid 'recomputes' and marking the field as modified (i.e. writing it to DB), you need to change the approach and use the ‘protecting’ method on the env. You could also improve readability and the batch processing by grouping your lines by invoice policy and using a context manager to apply the invoice policy to the products when computing the untaxed amount to invoice.

```python

@contextmanager
def _sale_invoice_policy(self, lines):
    """ Apply the sale invoice policy to the products
   
    This method must be called with lines sharing the same invoice policy
    """
    invoice_policy = lines.mapped("order_id.invoice_policy")
    if len(invoice_policy) > 1:
        raise Exception("The mehtod _sale_invoice_policy must be called with lines sharing the same invoice policy")
    invoice_policy = invoice_policy[0]
    invoice_policy_field = self.env["product.product"]._fields["invoice_policy"]
    products = lines.product_id
    with self.env.protecting([invoice_policy_field], products):
        old_values = {}
        for product in products:
            old_values[product] = product.invoice_policy
            product_id.invoice_policy = invoice_policy
        yield
        for product, invoice_policy in old_values.items():
            product.invoice_policy = invoice_policy

@api.depends("order_id.invoice_policy")
def _compute_untaxed_amount_to_invoice(self) -> None:
    other_lines = self.filtered(
        lambda line: line.product_id.type == "service"
        or line.order_id.invoice_policy == "product"
        or line.order_id.invoice_policy == line.product_id.invoice_policy
        or line.state not in ["sale", "done"]
    )
    super(SaleOrderLine, other_lines)._compute_untaxed_amount_to_invoice()
    for lines in (self - other_lines).partition("invoice_policy").values():
        with self._sale_invoice_policy(lines):
            super(SaleOrderLine, lines)._compute_untaxed_amount_to_invoice()
```

Reply:
Done